### PR TITLE
Refactor media services

### LINKED ISSUE #866: Simplify media services
## Problem

Currently, we have two almost identical services for media searching: [ImageService](https://github.com/WordPress/openverse-frontend/blob/c31a971088762d730375ccbe872677d91e4cdb57/src/data/image-service.js) and [AudioService](https://github.com/WordPress/openverse-frontend/blob/c31a971088762d730375ccbe872677d91e4cdb57/src/data/audio-service.js), both of which also use [BaseMediaService](https://github.com/WordPress/openverse-frontend/blob/c31a971088762d730375ccbe872677d91e4cdb57/src/data/base-media-service.js). 

## Description

Using one generic `MediaService` with a `mediaType` parameter would simplify the code and make it more maintainable, as well as simplify the addition of new media types in the future.

## Alternatives

An alternative would be to leave services separate. This would allow us to add methods that only exist for one media type to a single service. However, it would cause some inconveniences because:
- adding a new media type would require to create a new media service.
- a change in the API would require to change all services.

## Implementation

- [x] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #866 by @obulat

## Description

This PR joins three separate modules for `media-service`s into one generic `MediaService` that can be created with a `mediaType` parameter. All of the methods for both `audio` and `image` are the same. This change will allow us to simplify media fetching, and make the store conversion to Pinia and addition of other media types easier.

## Testing Instructions

Run the app as usual, and check that you can search for all media types. 

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by sarayourfriend** [comment]
This is great Olga! I haven't tested the code locally yet but I did take it for a spin to [do some fun TypeScript hoop-jumping](https://github.com/WordPress/openverse-frontend/pull/868) to make the MediaStore a generic class based on the `mediaType` in the constructor. Take it or leave it, up to you :slightly_smiling_face:

**Comment by obulat** [comment]
> This is great Olga! I haven't tested the code locally yet but I did take it for a spin to [do some fun TypeScript hoop-jumping](https://github.com/WordPress/openverse-frontend/pull/868) to make the MediaStore a generic class based on the `mediaType` in the constructor. Take it or leave it, up to you 🙂

Thank you, Sara, your PR is amazing! I've left my [notes there](https://github.com/WordPress/openverse-frontend/pull/868#pullrequestreview-882930040), and added some changes from it here.

File: src/data/api-service.js
Comment (sarayourfriend) [inline src/data/api-service.js:12]: The `*` (existential type) is basically `any` and will swallow the actual return type, which here is always `string`. Also, something to note, doing `AnyType | AWiderType` will always reduce to just `AWiderType`, so the `param` annotation will just reduce to `string`, which isn't as helpful. We can either leave out the `MediaType` annotation or update the call sites of this function to specify that the input is always `MediaType` (for example update type of `resource` on the `query` function to be `MediaType`).

```suggestion
 * @param {import('../constants/media').MediaType} resource
 * @returns {string}
```

File: src/store/types.d.ts
Comment (sarayourfriend) [inline src/store/types.d.ts:141]: These need to refer to a `FrontendMediaType` value.

```suggestion
    audio: MediaStoreResult<'audio'>
    image: MediaStoreResult<'image'>
```

**Review by sarayourfriend — APPROVED** [review]
LGTM! Tested locally and it works great. Left a couple more subtle TypeScript things but given the files aren't added to `tsconfig` yet anyway it's no big deal to leave them out, we'll just have to fix them later when we add them to `tsconfig`.

**Comment by sarayourfriend** [comment]
@WordPress/openverse-frontend can we get a second review on this? :slightly_smiling_face:

File: src/data/api-service.js
Comment (zackkrida) [inline src/data/api-service.js:17]: nice fix here.

**Review by zackkrida — APPROVED** [review]
The new abstraction is quite nice, and the types are fantastic. Some non-blocking questions:

- Should we add the `peaks` key here to the `AudioDetail` interface? Since it will be used for the waveforms soon and is already accessed in the `VWaveform` component implementations.
- How will we handle if a media type has a one-off endpoint, like the `/waveform` endpoint for audio or future endpoints that, say, 3D models might need?

**Comment by sarayourfriend** [comment]
> How will we handle if a media type has a one-off endpoint, like the /waveform endpoint for audio or future endpoints that, say, 3D models might need?

Subclass the `MediaService`:

```js
class Model3DService extends MediaService<'3d models'> {
  constructor() {
    super('3d models')
  }

  getCustomData() {
    // impl
  }
}
```

**Pull request merged**
